Add Giropay Payment Method

bborman22 opened the pull request:
Fixes #1510 

#### Changes proposed in this Pull Request
This PR is branched from the [Sepa PR](https://github.com/Automattic/woocommerce-payments/pull/1537) and shouldn't be merged until that one is merged to avoid conflicts. This PR adds the Giropay payment method. It is currently hidden behind a feature flag controlled by the `_wcpay_feature_giropay` option. This MVP implementation focuses on the successful payment flow using the classic editor. There will be more PRs to follow focusing on handling failed payment flows, block based checkout, and general optimization of the code.

**Edit:** This branch requires the payments server to be running the branch [here](740-gh-Automattic/woocommerce-payments-server) until it is merged. 

Some highlights to review:
* Allows the passing of additional fields to the API for creating an intent
* Adds the `next_action` and `last_payment_error` fields to the Payment Intention model
* Handles the `redirect_to_url` next action for intents with `requires_action` status
* Uses `wp` hook to check for the redirect back from Stripe and complete the order or error back to the checkout page
    * This hook was used in the [Stripe Gateway plugin](https://github.com/woocommerce/woocommerce-gateway-stripe), I didn't see any other obvious hook that fires early enough, but am open to other suggestions. 

### Testing instructions
* Set store currency to euro
* Add `_wcpay_feature_giropay` to `wp_options` table with a value of 1
* Enable Giropay by going to WooCommerce > Settings > Payments > Manage WooCommerce Payments (/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments) and using the checkbox at the bottom of the page.
* Add item to cart and go to checkout
* Choose Giropay and click Place Order
* You should be redirected to a Stripe page where you can view the payment information and choose to authorize or fail the payment.
* Choose Authorize and you should be redirected back to the order received page on the site.
* Check the Order in admin and see a successful payment and the proper payment notes recorded.
* Repeat the order process and choose fail to authorize the payment and you should be returned to the checkout page with an error displayed.

### Notes / Known Issues
[Stripe documentation](https://stripe.com/docs/payments/giropay/accept-a-payment#handle-redirect) used as reference
[Transaction details page for SEPA does not load in wcpay transaction tab](https://github.com/Automattic/woocommerce-payments/issues/1472 )


-------------------

- [ ] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

superdav42 commented on includes/payment-method/class-giropay.php:
```diff
@@ -0,0 +1,284 @@
+			wp_enqueue_script( 'wcpay-giropay-checkout' );
+
+			wp_enqueue_style(
+				'wcpay-giropay-checkout',
```
Probably we should use the same handle as card. `wcpay-checkout`. As it is we're embeding the same css file 3 times.
```
<link rel='stylesheet' id='wcpay-checkout-css'  href='https://dev.nnucomputerwhiz.com/plugins/woocommerce-payments/dist/checkout.css?ver=2.2.0' media='all' />
<link rel='stylesheet' id='wcpay-sepa-checkout-css'  href='https://dev.nnucomputerwhiz.com/plugins/woocommerce-payments/dist/checkout.css?ver=2.2.0' media='all' />
<link rel='stylesheet' id='wcpay-giropay-checkout-css'  href='https://dev.nnucomputerwhiz.com/plugins/woocommerce-payments/dist/checkout.css?ver=2.2.0' media='all' />
```
SEPA has the some problem. 

`wcpay_config` var is also duplicated. Probably we should pull these out into a separate protected method in the base class and call it it way so that it's only triggered once if 3 methods are enabled. We could put this optimization in a separate PR.

bborman22 replied:
Great, thanks for this! I renamed the style to avoid that duplication and [created this issue](https://github.com/Automattic/woocommerce-payments/issues/1598) for the `wcpay_config` optimization.

superdav42 commented on includes/payment-method/class-giropay.php:
```diff
@@ -0,0 +1,284 @@
+			];
+
+			return $this->process_payment_for_order( WC()->cart, $payment_information, $intent_api_parameters );
+		} catch ( Exception $e ) {
```
`use Exception;` is not in header so this is actually looking for \WCPay\Payment_Method\Exception which doesn't exist and will never be thrown. Alternatively you could change it to `\Exception $e` but I prefer to put it in the header.

bborman22 replied:
I added it to the header now, thanks!

superdav42 commented on includes/payment-method/class-giropay.php:
```diff
@@ -0,0 +1,284 @@
+use WC_Payments_Token_Service;
+use WC_Payments;
+use WC_Payments_Utils;
+
```
```suggestion
use Exception;

```

superdav42 commented on tests/unit/payment-method/test-class-giropay.php:
```diff
@@ -0,0 +1,217 @@
+use WP_User;
+
+/**
+ * WCPay\Payment_Gateway\Sepa Unit tests
```
```suggestion
 * WCPay\Payment_Gateway\Giropay Unit tests
```

superdav42 commented on tests/unit/payment-method/test-class-giropay.php:
```diff
@@ -0,0 +1,217 @@
+<?php
+/**
+ * Class Sepa_Test
```
```suggestion
 * Class Giropay_Test
```

superdav42 commented on tests/unit/payment-method/test-class-giropay.php:
```diff
@@ -0,0 +1,217 @@
+/**
+ * WCPay\Payment_Gateway\Sepa Unit tests
+ */
+class Giropay_Test extends WP_UnitTestCase {
```
Maybe in the future we can optimize these tests so they extend a base class and all these setup code will not need to be duplicated.

FangedParakeet replied:
I may attempt to do this when I shortly implement Sofort as a payment method to avoid duplicating code over and over.

superdav42 commented on client/checkout/classic/giropay.js:
```diff
@@ -0,0 +1,198 @@
+/* global jQuery */
```
We removed cliend/checkout/classic/sepa.js and put the sepa specific code into index.js along with the CC payment method code. I think we should do something similar with this giropay.js file. As it is several functions such as showError are duplicated.

superdav42 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1944,7 +1965,7 @@ public function get_transaction_url( $order ) {
 	 * @return string            Transaction details page url.
 	 */
-	private function compose_transaction_url( $charge_id ) {
+	public function compose_transaction_url( $charge_id ) {
```
```suggestion
	protected function compose_transaction_url( $charge_id ) {
```
Since \WCPay\Payment_Method\Giropay is a child of WC_Payment_Gateway_WCPay protected access should be enough and I don't think we want to make this public unless we have to.

superdav42 commented on includes/payment-method/class-giropay.php:
```diff
@@ -0,0 +1,284 @@
+	/**
+	 * Registers all scripts, necessary for the gateway.
+	 */
+	public function register_scripts() {
```
If we move code in giropay.js to index.js this whole method could be removed.

superdav42 commented on includes/payment-method/class-giropay.php:
```diff
@@ -0,0 +1,284 @@
+		$this->title              = __( 'Giropay', 'woocommerce-payments' );
+		$this->description        = __( 'You will be redirected to Giropay.', 'woocommerce-payments' );
+
+		add_action( 'wp', [ $this, 'maybe_process_redirect_order' ] );
```
I can't find any better hook

superdav42 reviewed (changes requested):
Thank for working on this and keeping up with the pattern I tried to set with SEPA. I'd like to see the js code be consolidated into index.js before we merge this, the rest are minor issues. Also I couldn't get the gateway to work on my site. Keeps giving the error:
Error: The provided PaymentMethod is already attached to another object. You cannot reuse PaymentMethods without attaching them to a Customer object first.

FangedParakeet reviewed (approved):
With the addition to the code on the server, this is working as expected for me! Code looks good and the flow makes sense to me. Pleasantly surprised and relieved that this entire flow could be implemented with almost no changes required on the server side.

I do agree with @superdav42 that `giropay.js` can be absorbed by `classic/index.js`. It is getting a little crowded in that file, but there is quite a lot of duplicated code in `giropay.js` and the minor differences in implementation do seem like they can reflected inside `index.js` with just a few minor adjustments. Other than Dave's other astute suggestions above, I don't really have any problems with anything else here. 

L G T M .

FangedParakeet commented:
> Keeps giving the error:
Error: The provided PaymentMethod is already attached to another object. You cannot reuse PaymentMethods without attaching them to a Customer object first.

@superdav42 I, too, reached this hurdle and promptly began falling onto my face: the missing magic step is this little snippet -> 740-gh-Automattic/woocommerce-payments-server that is required on the payments server.

bborman22 commented:
@FangedParakeet and @superdav42 Thank you both for taking a look! I believe I've addressed the issues mentioned in the latest commit. Let me know if there was anything else I could look into for this PR.

FangedParakeet commented:
Code changes look rosy to me. I just ran through a couple test checkouts--one authorizing the Giropay transaction and another failing it--and everything is still working as expected. It's got two thumbs up and a warm, well-wishing wink from me. 

L
G
T
M.

superdav42 commented:
With the branch on the payment server everything worked for me. This is G2G.
